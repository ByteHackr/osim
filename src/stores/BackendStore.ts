import { ref } from 'vue';

import { defineStore } from 'pinia';

export interface OsimBackends {
  error: boolean | string;
  isReady: () => boolean;
  osidb: string | unknown;
}

export const useBackendStore = defineStore('BackendStore', () => {
  const error = ref<string>('');
  const osidb = ref<string>('');

  function isReady() {
    return osidb.value !== '';
  }

  return {
    error,
    osidb,
    isReady,
  };
});
