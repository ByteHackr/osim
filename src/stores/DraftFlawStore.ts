import { ref } from 'vue';
import { defineStore } from 'pinia';
import type { ZodFlawType } from '@/types/zodFlaw';
import { unionWith } from 'ramda';
import type { ZodFlawAcknowledgmentType, ZodFlawReferenceType } from '@/types/zodFlaw';

export const useDraftFlawStore = defineStore('DraftFlawStore', () => {
  const draftFlaw = ref<ZodFlawType | null>(null);

  function saveDraftFlaw(flaw: ZodFlawType) {
    draftFlaw.value = flaw;
  }

  function addDraftFields(fetchedFlaw: ZodFlawType) {
    if (draftFlaw.value?.acknowledgments) {
      fetchedFlaw.acknowledgments = mergeAcks(fetchedFlaw.acknowledgments, draftFlaw.value.acknowledgments);
    }
    if (draftFlaw.value?.references) {
      fetchedFlaw.references = mergeRefs(fetchedFlaw.references, draftFlaw.value.references);
    }

    return fetchedFlaw;
  }

  const mergeRefs =
    (a: ZodFlawReferenceType[], b: ZodFlawReferenceType[]) => unionWith(refsComparator, a, b);

  const refsComparator = (a: ZodFlawReferenceType, b: ZodFlawReferenceType): boolean => {
    return a.url === b.url && a.description === b.description;
  };
  // If an element exists in both lists, the first element from the first list will be used.
  const mergeAcks =
    (a: ZodFlawAcknowledgmentType[], b: ZodFlawAcknowledgmentType[]) => unionWith(ackComparator, a, b);

  const ackComparator = (a: ZodFlawAcknowledgmentType, b: ZodFlawAcknowledgmentType): boolean => {
    return a.name === b.name && a.affiliation === b.affiliation;
  };

  function $reset() {
    draftFlaw.value = null;
  }

  return {
    $reset,
    draftFlaw,
    saveDraftFlaw,
    addDraftFields,
  };
});

