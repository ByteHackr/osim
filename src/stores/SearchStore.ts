import { computed } from 'vue';

import { defineStore } from 'pinia';
import { useLocalStorage } from '@vueuse/core';

export type SearchSchema = {
  queryFilter: string;
  searchFilters: Record<string, string>;
};

const defaultValues: SearchSchema = { searchFilters: {}, queryFilter: '' };

const _searchStoreKey = 'SearchStore';
const search = useLocalStorage(_searchStoreKey, defaultValues);
function saveFilter(filters: Record<string, string>, query: string = '') {
  search.value.searchFilters = filters;
  search.value.queryFilter = query;
}

function resetFilter() {
  search.value.searchFilters = {};
  search.value.queryFilter = '';
}
export const useSearchStore = defineStore('SearchStore', () => {
  const searchFilters = computed(() => search.value.searchFilters || {});
  const queryFilter = computed(() => search.value.queryFilter || '');

  return {
    searchFilters,
    queryFilter,
    saveFilter,
    resetFilter,
  };
});
