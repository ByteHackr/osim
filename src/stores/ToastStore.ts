import { ref } from 'vue';

import { defineStore } from 'pinia';
import { DateTime } from 'luxon';

import type { ToastProps } from '@/components/widgets/Toast.vue';

export interface ToastNew {
  body: string;
  bodyHtml?: boolean;
  css?: ToastProps['css'];
  timeoutMs?: number;
  title?: string;
}

interface ToastAdded extends ToastNew {
  id: number;
  isFresh: boolean;
  timestamp: DateTime;
}

let toastId = 0;

export const useToastStore = defineStore('ToastStore', () => {
  const toasts = ref<ToastAdded[]>([]);

  function addToast(toast: ToastNew) {
    const newToast: ToastAdded = {
      ...toast,
      id: toastId++,
      timestamp: DateTime.now(),
      isFresh: true,
    };
    toasts.value.unshift(newToast);
  }

  function $reset() {
    toasts.value = [];
  }

  return {
    $reset,
    addToast,
    toasts,
  };
});
