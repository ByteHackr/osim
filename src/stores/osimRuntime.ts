// No need to use Pinia; these values are set on app startup and never changed.
// Named with camelCase instead of PascalCase to differentiate from Pinia stores.

import {computed, ref} from 'vue';
import {z} from 'zod';

// export enum OsimRuntimeDev {
//   PROD,
//   DEV,
//   MOCK,
// }
// export const osimRuntimeDev = OsimRuntimeDev.DEV;

export enum OsimRuntimeStatus {
  INIT,
  READY,
  ERROR,
}

let status = ref(OsimRuntimeStatus.INIT);
export const osimRuntimeStatus = computed<OsimRuntimeStatus>(() => {
  return status.value;
});

const OsimRuntime = z.object({
  backends: z.object({
    osidb: z.string(),
  }),
  osimVersion: z.string(),
  error: z.string().default(''),
})
type OsimRuntime = z.infer<typeof OsimRuntime>;

let runtime = ref<OsimRuntime>({
  backends: {osidb: ''},
  osimVersion: '',
  error: '',
});
export const osimRuntime = computed<OsimRuntime>(() => {
  return runtime.value;
});

let setupCallCount = 0;

export function setup() {
  if (setupCallCount > 0) {
    console.warn('osimRuntime.setup called more than once:', setupCallCount + 1);
  }
  setupCallCount++;

  return fetch('/runtime.json')
      .then(response => response.json())
      .then((json: OsimRuntime) => {
        console.debug('OsimRuntime', json);
        try {
          runtime.value = OsimRuntime.parse(json);
          status.value = OsimRuntimeStatus.READY;
        } catch (e) {
          console.error('Unable to parse OsimRuntime', e);
          runtime.value.error = 'Backends are not correctly configured. Please try again later.';
          status.value = OsimRuntimeStatus.ERROR;
        }
      })
      .catch(e => {
        console.error('Unable to get backends', e);
        runtime.value.error = 'Error finding backends. Possible deployment misconfiguration. Please try again.';
        status.value = OsimRuntimeStatus.ERROR;
      });
}
