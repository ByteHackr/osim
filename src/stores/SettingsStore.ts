import { defineStore } from 'pinia';
import { z } from 'zod';
import { useLocalStorage } from '@vueuse/core';

import { useToastStore } from './ToastStore';

export const SettingsSchema = z.object({
  // bugzillaApiKey: z.string().length(
  //   32, {message: 'Bugzilla API Key is the wrong length!'}
  // ).optional(),
  bugzillaApiKey: z.string().optional().or(z.literal('')).default(''),
  jiraApiKey: z.string().optional().or(z.literal('')).default(''),
  showNotifications: z.boolean(),
  affectsPerPage: z.number(),
  trackersPerPage: z.number(),
  isHidingLabels: z.boolean().optional().default(false),
  privacyNoticeShown: z.boolean().default(false),
  singleCommentsView: z.boolean().default(false),
});

export type SettingsType = z.infer<typeof SettingsSchema>;

const defaultValues: SettingsType = {
  bugzillaApiKey: '',
  jiraApiKey: '',
  showNotifications: false,
  affectsPerPage: 10,
  trackersPerPage: 10,
  isHidingLabels: false,
  privacyNoticeShown: false,
  singleCommentsView: false,
};

export const useSettingsStore = defineStore('SettingsStore', () => {
  const { addToast } = useToastStore();

  const settings = useLocalStorage('OSIM::USER-SETTINGS', structuredClone(defaultValues));

  const validatedSettings = SettingsSchema.safeParse(settings.value);
  if (validatedSettings.success) {
    if (JSON.stringify(validatedSettings.data) !== JSON.stringify(settings.value)) {
      settings.value = validatedSettings.data;
    }
  } else {
    settings.value = structuredClone(defaultValues);
  }

  if (!settings.value.privacyNoticeShown) {
    addToast({
      title: 'Privacy Notice',
      body: 'OSIM transmits input information externally to Bugzilla for the purpose of retrieving bug data.' +
      ' In some cases that information may be publicly visible.',
    });
    settings.value.privacyNoticeShown = true;
  }

  function $reset() {
    settings.value = structuredClone(defaultValues);
  }

  return {
    $reset,
    settings,
  };
});
