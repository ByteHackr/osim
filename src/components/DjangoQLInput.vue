<script lang="ts" setup>
import { onMounted, onUnmounted, ref } from 'vue';

import DjangoQL from '@mrmarble/djangoql-completion';

import { osimRuntime } from '@/stores/osimRuntime';

const model = defineModel<string>();
const emit = defineEmits<{
  submit: [value: string];
}>();

const djangoCompletion = ref();
const textArea = ref<HTMLTextAreaElement | null>(null);

onMounted(() => {
  if (!textArea.value) {
    return;
  }

  const djangoQL = new DjangoQL({
    completionEnabled: true,
    introspections: `${osimRuntime.value.backends.osidb}/osidb/api/v1/introspection`,
    selector: textArea.value,
    autoResize: false,
    onSubmit: function (value: string) {
      model.value = value;
      emit('submit', value);
    },
  });

  djangoQL.popupCompletion();

  djangoCompletion.value = djangoQL;
});

onUnmounted(() => {
  if (djangoCompletion.value) {
    djangoCompletion.value.destroyCompletionElement();
  }
});
</script>

<template>
  <textarea
    ref="textArea"
    v-model="model"
    class="form-control"
  />
</template>
