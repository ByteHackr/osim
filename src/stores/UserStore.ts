import {ref, computed} from 'vue';
import {defineStore} from 'pinia';
import {z} from 'zod';
import jwtDecode from 'jwt-decode';
import type {JwtPayload} from 'jwt-decode';

import router from '@/router';
import {osimRuntime} from '@/stores/osimRuntime';

const _sessionStorageKey = 'UserStore';

const loginResponse = z.object({
  access: z.string(),
  refresh: z.string(),
  env: z.string(),
});

export const useUserStore = defineStore('UserStore', () => {


  const refresh = ref<string>('');

  const jwtRefresh = computed<JwtPayload | null>(() => {
    try {
      return jwtDecode(refresh.value);
    } catch (e) {
      console.debug('UserStore: refresh token not a valid JWT', refresh.value, e);
    }
    return null;
  });

  const userName = computed(() => {
    if (refresh.value === '') {
      return 'Not Logged In';
    }
    let sub = jwtRefresh.value?.sub;
    if (sub == null) {
      // @ts-ignore
      sub = 'User ID: ' + jwtAccess.value?.user_id;
    }
    if (sub == null) {
      // fallback: look for user in domain root
      sub = document.cookie
          .split(';')
          .map(x => x.trim())
          .filter(x => /^.._user=/.test(x))
          .map(x => x.substring('.._user='.length))
          .reduce((acc: string[], x) => acc.concat(x.split('|')), [])
          .map(x => decodeURIComponent(x))
          .find(x => true)
    }
    if (sub == null) {
      sub = 'Current User';
    }
    return sub;
  });

  const env = ref<string>(''); // Placeholder: environment

  // let refreshInterval = 30000;
  // let refreshId = 0;


  function setTokens(refresh_: string, env_: string) {
    refresh.value = refresh_;
    env.value = env_;
    // refreshId = setInterval(performTokenRefresh, 5000);
    // writeSessionStorage()
  }

  readSessionStorage(); // Initially read when loading UserStore
  function readSessionStorage() {
    let storedUserStore = sessionStorage.getItem(_sessionStorageKey);
    if (storedUserStore != null) {
      try {
        storedUserStore = JSON.parse(storedUserStore);
        // @ts-ignore
        access.value = storedUserStore.access;
        // @ts-ignore
        refresh.value = storedUserStore.refresh;
        // @ts-ignore
        env.value = storedUserStore.env;
      } catch (e) {
        console.error('UserStore: unable to restore from sessionStorage', e);
        $reset();
      }
    }
  }

  // function login(jwtAccess, jwtRefresh) {
  //   return userService.login()
  //       .then(user => {
  //         jwt.value.user = JSON.stringify(user);
  //       })
  // }
  async function login() {
    // return fetch('https://osidb-stage.prodsec.redhat.com/auth/token', {
    return fetch(`${osimRuntime.value.backends.osidb}/auth/token`, {
      // credentials: 'same-origin',
      credentials: 'include',
      cache: 'no-cache',
    })
        .then(response => response.json())
        .then(json => {
          const parsedLoginResponse = loginResponse.passthrough().parse(json);
          access.value = parsedLoginResponse.access;
          refresh.value = parsedLoginResponse.refresh;
          env.value = parsedLoginResponse.env;
        })
        .catch(e => {
          console.error('UserStore: unsuccessful login request', e);
        });
  }

  // function writeSessionStorage() {
  //   let storedUserStore = {
  //     access: access.value,
  //     refresh: refresh.value,
  //     env: _env.value,
  //     _modifyDate: Date.now(),
  //   };
  //   sessionStorage.setItem(_sessionStorageKey, JSON.stringify(storedUserStore));
  // }

  function logout() {
    // clearInterval(refreshInterval);
    $reset();

    // router.push({name: 'login'});
    // return Promise.resolve()
    //     .then(() => {
    //       jwt.value.user = undefined;
    //     })
    //     .then(() => {
    //       router.push({name: 'login'})
    //     })
    return router.push({name: 'login'});
  }

  /**
   * Side effect: wipes tokens if tokens are expired
   */
  function isAuthenticated() {
    let now = Date.now();
    let refreshExp = jwtRefresh.value?.exp;
    if (refreshExp != null) {
      return now < refreshExp * 1000;
    }
    $reset();
    return false;
  }

  function $reset() {
    setTokens('', '');
  }

  return {
    refresh,
    jwtRefresh,
    userName,
    env,
    login,
    logout,
    isAuthenticated,
    $reset,
  };
});

