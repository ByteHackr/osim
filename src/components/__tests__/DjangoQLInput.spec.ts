import { mount } from '@vue/test-utils';
import { http, HttpResponse } from 'msw';

import DjangoQLInput from '@/components/DjangoQLInput.vue';

import { osimRuntime } from '@/stores/osimRuntime';
import { server } from '@/__tests__/setup';

describe('djangoQLInput', () => {
  beforeAll(() => {
    server.use(
      http.get(`${osimRuntime.value.backends.osidb}/osidb/api/v1/introspection`,
        () => HttpResponse.json({}, { status: 200 })),
    );
    vi.useFakeTimers();
  });

  it('should render', async () => {
    const wrapper = mount(DjangoQLInput);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should instantiate djangoql-completion on mount', async () => {
    mount(DjangoQLInput);

    expect(document.body.querySelector('.djangoql-completion')).toBeTruthy();
  });

  it('should cleanup on unmount', async () => {
    const wrapper = mount(DjangoQLInput);
    wrapper.unmount();

    expect(document.body.querySelector('.djangoql-completion')).toBeFalsy();
  });

  it('should emit `submit` event on suggestions submit', async () => {
    const wrapper = mount(DjangoQLInput);

    await wrapper.find('textarea').setValue('impact = "Low"');
    await wrapper.find('textarea').trigger('keydown.Enter');

    expect(wrapper.emitted()).toHaveProperty('submit');
    expect(wrapper.emitted('submit')![0]).toEqual(['impact = "Low"']);
  });
});
