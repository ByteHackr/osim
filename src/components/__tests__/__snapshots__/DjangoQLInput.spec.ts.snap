// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`djangoQLInput > should render 1`] = `"<textarea class="form-control" autocomplete="off"></textarea>"`;
